lib/expand/join.go

```diff
@@ -469,66 +469,107 @@ func (p *Peer) agentURL(ctx operationContext) (string, error) {
 }
 
-func (p *Peer) run() error {
-	ctx, err := p.connect()
+// getAgent creates an RPC agent instance that, once started, will connect
+// to its peer which can be either installer process or existing cluster
+func (p *Peer) getAgent(opCtx operationContext) (*rpcserver.PeerServer, error) {
+	agentURL, err := p.agentURL(opCtx)
 	if err != nil {
-		return trace.Wrap(err)
+		return nil, trace.Wrap(err)
 	}
-
-	err = p.ensureServiceUserAndBinary(*ctx)
+	listener, err := net.Listen("tcp", defaults.GravityRPCAgentAddr(p.AdvertiseAddr))
 	if err != nil {
-		return trace.Wrap(err)
+		return nil, trace.Wrap(err)
+	}
+	defer func() {
+		if err != nil {
+			listener.Close()
+		}
+	}()
+	agent, err := install.StartAgent(agentURL, rpcserver.PeerConfig{
+		Config: rpcserver.Config{
+			Listener:      listener,
+			Credentials:   opCtx.Creds,
+			RuntimeConfig: p.RuntimeConfig,
+		},
+		WatchCh: p.WatchCh,
+		ReconnectStrategy: rpcserver.ReconnectStrategy{
+			ShouldReconnect: utils.ShouldReconnectPeer,
+		},
+	}, p.WithField("peer", listener.Addr().String()))
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	// make sure that connection to the RPC server can be established
+	ctx, cancel := context.WithTimeout(p.Context, defaults.PeerConnectTimeout)
+	defer cancel()
+	err = agent.Check(ctx)
+	if err != nil {
+		return agent, trace.Wrap(err)
 	}
+	return agent, nil
+}
 
-	err = p.checkAndSetServerProfile(ctx.Cluster.App)
+func (p *Peer) run() error {
+	ctx, err := p.connect()
 	if err != nil {
 		return trace.Wrap(err)
 	}
 
-	agentURL, err := p.agentURL(*ctx)
+	err = p.ensureServiceUserAndBinary(*ctx)
 	if err != nil {
 		return trace.Wrap(err)
 	}
 
-	listener, err := net.Listen("tcp", defaults.GravityRPCAgentAddr(p.AdvertiseAddr))
+	err = p.checkAndSetServerProfile(ctx.Cluster.App)
 	if err != nil {
 		return trace.Wrap(err)
 	}
 
-	config := rpcserver.PeerConfig{
-		Config: rpcserver.Config{
-			Listener:      listener,
-			Credentials:   ctx.Creds,
-			RuntimeConfig: p.RuntimeConfig,
-		},
-		WatchCh: p.WatchCh,
-		ReconnectStrategy: rpcserver.ReconnectStrategy{
-			ShouldReconnect: utils.ShouldReconnectPeer,
-		},
-	}
+	// schedule cleanup function in case anything goes wrong before
+	// the operation can start
+	defer func() {
+		if err == nil {
+			return
+		}
+		p.Warnf("Peer is exiting with error: %v.", trace.DebugReport(err))
+		stopCtx, cancel := context.WithTimeout(p.Context, defaults.AgentStopTimeout)
+		defer cancel()
+		p.Warn("Stopping peer.")
+		if err := p.Stop(stopCtx); err != nil {
+			p.Errorf("Failed to stop peer: %v.", trace.DebugReport(err))
+		}
+		// in case of join via CLI the operation has already been created
+		// above but the agent failed to connect so we're deleting the
+		// operation because from user's perspective it hasn't started
+		//
+		// in case of join via UI the peer is joining to the existing
+		// operation created via UI so we're not touching it and the
+		// user can cancel it in the UI
```

Comment:
This means inconsistent handling:
  - if started (or created?) in UI, user is responsible for cleaning up the operation (cancelling, which I suppose, will mark it as failed)
  - if created in CLI flow, the operation vanishes completely even though it was created in the db and there was a failure to progress it

whereas if it were always moved to a failed state (for UI flow as well) uniformly, you need not bother with this special handling.

Reply:
On the contrary, it is consistent with the user expectations. If they created the operation in the UI they likely don't want it to suddenly disappear or fail if the agent failed to connect.